✨ feat: 새로운 chat, notification 기능 추가

hongdyuk opened the pull request:
본문
- 이유 chat/views.py, serializers.py, urls.py : 채팅방 목록 조회 및 생성 api, 특정 채팅방 조회, 수정 및 삭제 api, 메세지 목록 조회 및 생성 api, url 연결, serializer로 유효성 검사
- 이유 2 notification/views.py, serializers.py, urls.py, routing.py, consumers.py : 해당 유저에 대한 알림 목록 조회 및 읽지 않은 알림의 수를 같이 반환, 특정 알림 읽음 처리 전체 알림 읽음 처리, 알림 유효성 검사, url연결, routing웹소켓설정, asgi.py channels활성화

### 📌 관련 이슈
- 이슈 번호: # (연관된 이슈 번호를 적어주세요)

### 📝 작업 내용
- [x] 변경 사항 설명 1
- [x] 변경 사항 설명 2
- [ ] 기타 필요한 설명

### 📋 체크리스트
- [ ] 코드에 오류가 없는지 확인했습니다.
- [ ] 문서가 업데이트되었습니다.
- [ ] 테스트를 통과했습니다.

### 🔗 참고 자료
- 추가로 참고할 자료가 있다면 링크를 추가하세요.

### 아래 # 뒤에 PR머지와 함께 Close할 이슈번호를 입력해주세요.
closes #

hongdyuk committed:
✨ feat: 새로운 chat, notification 기능 추가

본문
- 이유 chat/views.py, serializers.py, urls.py : 채팅방 목록 조회 및 생성 api, 특정 채팅방 조회, 수정 및 삭제 api, 메세지 목록 조회 및 생성 api, url 연결, serializer로 유효성 검사
- 이유 2 notification/views.py, serializers.py, urls.py, routing.py, consumers.py : 해당 유저에 대한 알림 목록 조회 및 읽지 않은 알림의 수를 같이 반환, 특정 알림 읽음 처리 전체 알림 읽음 처리, 알림 유효성 검사, url연결, routing웹소켓설정, asgi.py channels활성화

hongdyuk committed:
✨ feat: 새로운 chat, notification 기능 추가

본문

이유 chat/views.py, serializers.py, urls.py : 채팅방 목록 조회 및 생성 api, 특정 채팅방 조회, 수정 및 삭제 api, 메세지 목록 조회 및 생성 api, url 연결, serializer로 유효성 검사
이유 2 notification/views.py, serializers.py, urls.py, routing.py, consumers.py : 해당 유저에 대한 알림 목록 조회 및 읽지 않은 알림의 수를 같이 반환, 특정 알림 읽음 처리 전체 알림 읽음 처리, 알림 유효성 검사, url연결, routing웹소켓설정, asgi.py channels활성화

화

Meoyoug commented on src/chat/serializers/chat_serializers.py:
12 - 13 라인에서 user와 expert 객체를 각각 밸리데이션해서 각각 표시되도록 해주세요.
또는 validate_field명 메서드를 사용해서 각각의 필드마다 밸리데이션을 할 수도 있어용
만약 유저가 없으면
"유저의 데이터는 필수입니다." 

전문가가 없으면
"전문가의 데이터는 필수입니다."

Meoyoug commented on src/chat/serializers/chat_serializers.py:
26번째 줄에 content가 없으면 이라는 조건 뒤에 or not data["content"].strip()을 추가조건으로 한 이유가 뭘까요?

Meoyoug commented on src/chat/views/chat_views.py:
get_queryset 메서드에서 요청을 보낸 유저와 상관없이 EstimationRequest의 상태를 살펴서 가져오는 것 같은데 확인해보셔야할 거 같아요 해당 API는 요청을 보낸 유저가 자신의 채팅방을 조회하는 건데 본인과 관계없는 채팅방도 다 가져오도록 되어있네요.

Meoyoug replied:
ChatRoom이 EstimationRequest를 참조하지않는데 filter가 적용이되나요?

Meoyoug commented on src/notifications/consumers.py:
text_data에 알림의 메시지만 전달하고 있네요
알림에 대한 추가적으로 사용될 수 있는 정보를 다 전달해주세요.
그리고 컨슈머만 있다고해서 알림을 전달할 수 있는게 아니라 장고의 signal을 사용해서 특정 개체의 db execute가 일어날 때 알림객체를 생성하고 컨슈머로 전달해야합니다.

Meoyoug commented on src/notifications/serializers.py:
제목은 50자 정도로 풀고 메시지는 TEXT 필드로 바꿔야할 거 같네요 너무 제한적이에요

Meoyoug replied:
여기서 update 메소드를 오버라이드해서 읽음처리를 구현해주세요

Meoyoug replied:
이거에 대한 모델 변경은요?

Meoyoug commented on src/notifications/views.py:
def patch(self, request, *args, **kwargs):
        notification = self.get_object()
        notification.is_read = True
        notification.save()
        serializer = self.get_serializer(notification)
        channel_layer = get_channel_layer()
        async_to_sync(channel_layer.group_send)(
            f"notification_{self.request.user.id}",
            {"type": "send_notification", "message": f"알림 {notification.id}이(가) 읽음 처리되었습니다."},
        )
        return Response(serializer.data)
        
 이부분에서 serializer = self.get_serializer(notification) 후에 serializer.is_valid 메서드를 호출하고 is_read 필드를 시리얼라이저의 update 메서드를 통해 변경되도록 바꿔주세요.
그리고 읽음처리에 대한 응답은 웹소켓으로 전달하지않고 api응답으로만 전달되어도 됩니다.
response에서는 serializer.data와 함께 status=status.200_OK 이런식으로 명시해주세요.

Meoyoug reviewed (commented):
전체적으로 수정할 부분이 많아 보이네요 코멘트 달아 놓은 부분 수정 또는 추가해주시구요
Response의 data에 해당하는 부분에 메시지를 보낼때는 "detail": "요청이 처리되었습니다." 이런식으로 detail 키값에 메시지를 담아주세요.

hongdyuk committed:
💡 chore: 각 api의 코드 수정 및 생성

본문
1. chat/view, serializer 코드 수정 및 보완 작업 완료
2. notification/view, serializer, consmers 코드 수정 및 보완 작업 완료

hongdyuk committed:
💡 chore: 이전 commit 수정 및 보완

본문
- 이유 chat, notification pr 코멘트 수정 및 보완

Meoyoug commented on src/notifications/consumers.py:
각 앱에 signals.py를 따로 만들어서 거기에다가 각각의 모델이 생성될 때 signal 사용해서 Notification 객체를 생성하고 Notification 객체가 생성되면 여기 정의하신 리시버로 알림을 웹소켓으로 보내도록 해주세요.

hongdyuk committed:
💡 chore: test 코드 삭제

- 이유 : test 코드 삭제

hongdyuk committed:
💡 chore: 머지 수정

- 이유 : git merge 수정

hongdyuk committed:
💡 chore: 머지 컴플릿 수정

본문
- 이유 :  scripts/generate_dummy.sh
        src/chat/serializers/chat_serializers.py
        src/chat/tests/tests.py
        src/common/signals/__init__.py
        src/common/signals/chat_signals.py
        src/common/signals/estimation_signals.py
        src/common/signals/notification_signals.py
        src/common/signals/request_signals.py
        src/common/signals/reservation_signals.py
        src/common/signals/review_signals.py
        src/estimations/tests.py
         Dockerfile
        docker-compose.yml
        scripts/entrypoint.sh
        src/chat/views/chat_views.py
        src/common/constants/choices.py
        src/config/settings/base.py
        src/notifications/consumers.py
        src/notifications/tests.py
        src/notifications/views.py

hongdyuk committed:
test.sh 가동

hongdyuk committed:
kubernetes 파일 백업

hongdyuk committed:
💡 chore: chat api 수정 및 보완 + 테스트 코드 작성, websocket 수정 및 보완 + 테스트 코드 작성

본문
- 이유 1 : chat/ chat_serializers.py, chat_views.py, models.py, test_views.py, test_consumer.py 수정 및 작성
- 이유 2 : config/ asgi.py, base.py, routing.py 수정 및 추가

Meoyoug reviewed (approved):
고생하셨습니다
